### PR TITLE
Add ROS support

### PR DESCRIPTION
**Pull request by gizatt**
At last, here's a draft for how I propose we incorporate a catkin workspace into the build.

The main thing I want make sure we agree on first is just where the catkin workspace lives -- it works nicely as a subdirectory of `src`, where everything inside is a Catkin package, and everything above can just be whatever it wants. We could probably shift that up a directory (and have `src` just be a catkin directory), and ensure that everything already-existing doesn't break Catkin (i.e. either is a real catkin package, or doesn't have a `package.xml` at all). It'll be dirtier, in that the `CMakeLists.txt` file will have to dispatch Catkin for the compatible subdirectories, and ExternalProject for the rest. Not too awful, but it'd be a kinda clunky `CMakeLists.txt` that is half-automated and half-hand-curated.

The way that ROS actually gets used in the Docker image is still pretty clunky -- I haven't focused on cleaning that up yet. Making the ROS installation optional is a possibility, though it'll require piping more options through Docker. Not too bad to do. Also, I haven't tried basing the docker image on the ros-kinetic-xenial-~~~ image that I know is out there -- I can experiment with that before this is merged if y'all think it's a high priority.

I included a really simple perception demo to show that the Catkin build works, but it probably won't work on other machines yet (largely because it only works for Realsense R200's, and udev rules for cameras are still clunky).

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/158)

### CODE REVIEW

**Comment by manuelli** [comment]
Fixes #152

**Comment by patmarion** [comment]
btw, there is a catkin_ws in director, provided as an example.  you don't need to build it for spartan, but if it's easy then it might be nice.  normally i just cd to the director/catkin_ws and catkin make.

https://github.com/RobotLocomotion/director/blob/master/catkin_ws

**Comment by gizatt** [comment]
Jenkins please test

(want to see that the queue works)

**Comment by gizatt** [comment]
@patmarion Noted! I don't think I'll try to get that into this PR, but I'll see if I can work out a way to do it. I kinda cheated the Spartan catkin-ws into working by treating it as an external project and avoiding ever actually invoking catkin-build, but I think a carefully crafted external project for the Director catkin ws would work... will make an issue and see what I can do (probably as soon as we start using Director with ros).

**Comment by gizatt** [comment]
Jenkins please test

**Comment by gizatt** [comment]
Jenkins please test

**Comment by gizatt** [comment]
Jenkins please test

**Comment by gizatt** [comment]
Jenkins please test

File: config/setup_environment.sh.in [hunk 1]
```diff
@@ -13,8 +13,14 @@ export LD_LIBRARY_PATH=$SPARTAN_INSTALL_DIR/lib:$LD_LIBRARY_PATH
...
+# Some of these that we use live in drake's manipulation folder:
+export ROS_PACKAGE_PATH=$ROS_PACKAGE_PATH:$SPARTAN_SOURCE_DIR/drake/drake/manipulation/models
+
 alias makedirector='make -C $SPARTAN_BUILD_DIR/director/src/director-build install'
 alias makedrake='make -C $SPARTAN_BUILD_DIR/drake/drake install'
 
 alias cds='cd $SPARTAN_SOURCE_DIR'
 alias cdbuild='cd $SPARTAN_BUILD_DIR'
+
+source $SPARTAN_INSTALL_DIR/setup.sh
```
Comment (manuelli) [inline config/setup_environment.sh.in]: what is this `setup.sh` file?

**Comment by manuelli** [comment]
Other than the one minor above comment about `setup.sh` LGTM.

File: config/setup_environment.sh.in (hunk 1 above)
Comment (gizatt) [reply config/setup_environment.sh.in]: This is the environment setup file for the installed spartan catkin ws -- same thing that would normally be at `catkin_ws/install/setup.sh`. Sourcing this allows ROS to find Spartan's packages.

**Comment by manuelli** [comment]
Review status: 0 of 44 files reviewed at latest revision, 1 unresolved discussion.

---

*[config/setup_environment.sh.in, line 26 at r6](https://reviewable.io:443/reviews/robotlocomotion/spartan/158#-KwM3kure_fgfIrBdWAZ-r6-26:-KwXSTblm9ah6TbEeJKK:byum8hh) ([raw file](https://github.com/robotlocomotion/spartan/blob/4888af87056b0a8fd3c4a483dd162801a3965e14/config/setup_environment.sh.in#L26)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

This is the environment setup file for the installed spartan catkin ws -- same thing that would normally be at `catkin_ws/install/setup.sh`. Sourcing this allows ROS to find Spartan's packages.
</blockquote></details>

So I thought you wanted to source `catkin_ws/devel/setup.bash`, I'm not sure what the difference between `devel` and `install` is for `catkin` builds. Anyways just following the tutorial [here](http://wiki.ros.org/catkin/Tutorials/create_a_workspace) 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/158)*

**Comment by gizatt** [comment]
Review status: 0 of 44 files reviewed at latest revision, 1 unresolved discussion.

---

*[config/setup_environment.sh.in, line 26 at r6](https://reviewable.io:443/reviews/robotlocomotion/spartan/158#-KwM3kure_fgfIrBdWAZ-r6-26:-KwXXJmDm-L22clur79c:b5757mu) ([raw file](https://github.com/robotlocomotion/spartan/blob/4888af87056b0a8fd3c4a483dd162801a3965e14/config/setup_environment.sh.in#L26)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

So I thought you wanted to source `catkin_ws/devel/setup.bash`, I'm not sure what the difference between `devel` and `install` is for `catkin` builds. Anyways just following the tutorial [here](http://wiki.ros.org/catkin/Tutorials/create_a_workspace) 
</blockquote></details>

I think the only difference is that `install` will only include files that you explicitly wrote install rules for (just like Drake only exposes files in `build/install/***` that are explicitly installed), whereas `devel` includes all files that were built by the project (but not necessary installed / meant to be externally-visible). I think the `install` route is cleaner, given that we're treating these catkin packages as external projects (to a degree)... but I think the Draper code used `devel/setup.sh` instead, as I needed to add install rules for everything I added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/158)*

**Comment by manuelli** [comment]
Review status: 0 of 44 files reviewed at latest revision, 1 unresolved discussion.

---

*[config/setup_environment.sh.in, line 26 at r6](https://reviewable.io:443/reviews/robotlocomotion/spartan/158#-KwM3kure_fgfIrBdWAZ-r6-26:-KwXnea22UpLv1LskzzA:b-7qtonr) ([raw file](https://github.com/robotlocomotion/spartan/blob/4888af87056b0a8fd3c4a483dd162801a3965e14/config/setup_environment.sh.in#L26)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

I think the only difference is that `install` will only include files that you explicitly wrote install rules for (just like Drake only exposes files in `build/install/***` that are explicitly installed), whereas `devel` includes all files that were built by the project (but not necessary installed / meant to be externally-visible). I think the `install` route is cleaner, given that we're treating these catkin packages as external projects (to a degree)... but I think the Draper code used `devel/setup.sh` instead, as I needed to add install rules for everything I added.
</blockquote></details>

Ok what file is specifying how this copy is happening? I grep'd around the diffs in your PR but couldn't figure it out. Is there a reason why we don't just leave it in `$SPARTAN_INSTALL_DIR/catkin_project/install/setup.sh` and add that line to the `setup_environment.sh` file?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/158)*

**Comment by gizatt** [comment]
Review status: 0 of 44 files reviewed at latest revision, 1 unresolved discussion.

---

*[config/setup_environment.sh.in, line 26 at r6](https://reviewable.io:443/reviews/robotlocomotion/spartan/158#-KwM3kure_fgfIrBdWAZ-r6-26:-KwaNUc394V9VhcE646T:busq6xs) ([raw file](https://github.com/robotlocomotion/spartan/blob/4888af87056b0a8fd3c4a483dd162801a3965e14/config/setup_environment.sh.in#L26)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

Ok what file is specifying how this copy is happening? I grep'd around the diffs in your PR but couldn't figure it out. Is there a reason why we don't just leave it in `$SPARTAN_INSTALL_DIR/catkin_project/install/setup.sh` and add that line to the `setup_environment.sh` file?
</blockquote></details>

The copy of the catkin package install files (including `setup.sh` into the root of the install folder at CMAKE_INSTALL_PREFIX) is an automatic thing that Catkin does given that we're using `$CMAKE_INSTALL_PREFIX` as the install path (so that the packages can install regular executables, libraries, etc if they want to and  have them be visible to the rest of the project). Catkin just wants to put `setup.sh` at the root of that install directory.

I think our choice is between having `setup.sh` installed somewhere more precisely named (or I guess I could write a special install rule to rename or move it...), or having all catkin packages installed to a special catkin install directory (which other projects would not be able to link against by default).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/158)*

**Comment by manuelli** [comment]
Review status: 0 of 44 files reviewed at latest revision, 1 unresolved discussion.

---

*[config/setup_environment.sh.in, line 26 at r6](https://reviewable.io:443/reviews/robotlocomotion/spartan/158#-KwM3kure_fgfIrBdWAZ-r6-26:-Kwa_5rN1A5XT3QQ1nsQ:bongoz2) ([raw file](https://github.com/robotlocomotion/spartan/blob/4888af87056b0a8fd3c4a483dd162801a3965e14/config/setup_environment.sh.in#L26)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

The copy of the catkin package install files (including `setup.sh` into the root of the install folder at CMAKE_INSTALL_PREFIX) is an automatic thing that Catkin does given that we're using `$CMAKE_INSTALL_PREFIX` as the install path (so that the packages can install regular executables, libraries, etc if they want to and  have them be visible to the rest of the project). Catkin just wants to put `setup.sh` at the root of that install directory.

I think our choice is between having `setup.sh` installed somewhere more precisely named (or I guess I could write a special install rule to rename or move it...), or having all catkin packages installed to a special catkin install directory (which other projects would not be able to link against by default).
</blockquote></details>

Ok thanks for the explanation. I think it's fine for now, we can revisit it later as needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/158)*

**Pull request merged**